cpu/native/socket_zep/socket_zep.c

```diff
@@ -439,128 +569,47 @@ int _len(ieee802154_dev_t *dev)
-    real_send(zepdev->sock_fd, &hdr, sizeof(hdr), MSG_MORE);
-    real_send(zepdev->sock_fd, ack, sizeof(ack), MSG_MORE);
-    real_send(zepdev->sock_fd, &chksum, sizeof(chksum), 0);
+    return hdr->length - 2;
 }
 
 static int _read(ieee802154_dev_t *dev, void *buf, size_t max_size,
                  ieee802154_rx_info_t *info)
 {
-    int res;
     socket_zep_t *zepdev = dev->priv;
-    size_t frame_len = max_size + sizeof(zep_v2_data_hdr_t) + 2;
+    int res = 0;
 
     DEBUG("socket_zep::read: reading up to %zu bytes into %p\n", max_size, buf);
 
-    if (frame_len > sizeof(zepdev->rcv_buf)) {
-        DEBUG("socket_zep::read: frame size (%zu) exceeds RX  buffer (%zu bytes)\n",
-              frame_len, sizeof(zepdev->rcv_buf));
-        res = -ENOBUFS;
+    if (buf == NULL || zepdev->rcv_len == 0) {
         goto out;
     }
 
-    res = real_recv(zepdev->sock_fd, zepdev->rcv_buf, frame_len, MSG_TRUNC);
-
-    DEBUG("socket_zep::read: got %d/%zu bytes\n", res, frame_len);
+    DEBUG("socket_zep::read: %zu/%zu bytes into %p\n",
+           max_size, zepdev->rcv_len - sizeof(zep_v2_data_hdr_t) - 2, buf);
 
-    if (res <= (int)sizeof(zep_v2_data_hdr_t) || res > (int)frame_len) {
-        DEBUG("socket_zep::read: %s\n", strerror(errno));
-        res = 0;
-        goto out;
-    }
-
-    zep_hdr_t *tmp = (zep_hdr_t *)zepdev->rcv_buf;
-
-    if ((tmp->preamble[0] != 'E') || (tmp->preamble[1] != 'X')) {
-        DEBUG("socket_zep::read: invalid ZEP header\n");
+    if (max_size != zepdev->rcv_len - sizeof(zep_v2_data_hdr_t) - 2) {
+        DEBUG("socket_zep::read: size mismatch!\n");
         res = -EINVAL;
         goto out;
     }
 
-    if (tmp->version != 2) {
-        DEBUG("socket_zep::read: unsupported ZEP version %u\n", tmp->version);
-        res = -EINVAL;
-        goto out;
-    }
-
-    switch (((zep_v2_ack_hdr_t *)tmp)->type) {
-    case ZEP_V2_TYPE_DATA: {
-        zep_v2_data_hdr_t *zep = (zep_v2_data_hdr_t *)tmp;
+    zep_v2_data_hdr_t *zep = (zep_v2_data_hdr_t *)zepdev->rcv_buf;
 
-        if (zep->chan != zepdev->chan) {
-            DEBUG("socket_zep::read: wrong channel\n");
-            res = -EINVAL;
-            break;
-        }
-
-        if (info) {
-            info->lqi = zep->lqi_val;
-            info->rssi = -IEEE802154_RADIO_RSSI_OFFSET;
-        }
-
-        if (_dst_not_me(zepdev, zep + 1)) {
-            DEBUG("socket_zep::read: dst not me\n");
-            res = -EINVAL;
-            break;
-        }
-
-        _send_ack(zepdev, zep + 1);
-
-        memcpy(buf, zep + 1, max_size);
-        res = max_size;
-
-        break;
-    }
-    default:
-        DEBUG("socket_zep::read: unknown type %u\n", ((zep_v2_ack_hdr_t *)tmp)->type);
-        res = -EINVAL;
-        break;
+    if (info) {
+        info->lqi = zep->lqi_val;
+        info->rssi = lqi_to_rssi[zep->lqi_val]
+                   + ((random_uint32() & 0x3) - 2);
     }
 
-out:
-    _continue_reading(zepdev);
+    memcpy(buf, zep + 1, zep->length - 2);
```

Comment:
Could you explain this memcpy to me?

Reply:
We skip the header (`zep + 1`) and the frame checksum at the end (`zep->length - 2`)

Reply:
Would you mind adding an explaining comment in the code? :)

Reply:
[Sure](https://github.com/RIOT-OS/RIOT/compare/1585c8267891c0e437d44d7a3c831d9f5e51cd2e..0f965442d15e7277b7c6c4de4ab154966e30a3b6)!